tapdb/sqlc/migrations/000020_asset_unique_key.up.sql

```diff
@@ -1,4 +1,105 @@
+CREATE TABLE tmp_min_assets AS
+SELECT MIN(asset_id) AS min_asset_id,
+       genesis_id,
+       script_key_id,
+       amount,
+       anchor_utxo_id,
+       spent
+FROM assets
+GROUP BY genesis_id, script_key_id, amount, anchor_utxo_id, spent;
+
+-- Step 3: Create a mapping table to track old and new asset_ids.
+CREATE TABLE tmp_asset_id_mapping AS
+SELECT a.asset_id       AS old_asset_id,
+       tmp.min_asset_id AS new_asset_id
+FROM assets a
+         JOIN tmp_min_assets tmp
+              ON a.genesis_id = tmp.genesis_id
+                  AND a.script_key_id = tmp.script_key_id
+                  AND a.amount = tmp.amount
+                  AND a.anchor_utxo_id = tmp.anchor_utxo_id
+                  AND a.spent = tmp.spent;
+
+-- Step 4: To make the next step possible, we need to disable a unique index on
+-- the asset_witnesses table. We'll re-create it later.
+DROP INDEX IF EXISTS asset_witnesses_asset_id_witness_index_unique;
+
+-- Step 5: Update the asset_witnesses and asset_proofs tables to reference the
+-- new asset_ids.
+UPDATE asset_witnesses
+SET asset_id = (SELECT new_asset_id
+                FROM tmp_asset_id_mapping
+                WHERE old_asset_id = asset_witnesses.asset_id)
+WHERE EXISTS (SELECT 1
+              FROM tmp_asset_id_mapping
+              WHERE old_asset_id = asset_witnesses.asset_id);
+
+-- For the proofs we need skip re-assigning them to the asset that we're going
+-- to keep if it already has a proof. This is because the unique index on the
+-- asset_proofs table would prevent us from doing so. And we can't disable the
+-- unique index, because it is an unnamed/inline index.
+UPDATE asset_proofs
+SET asset_id = (SELECT new_asset_id
+                FROM tmp_asset_id_mapping
+                WHERE old_asset_id = asset_proofs.asset_id)
+WHERE EXISTS (SELECT 1
+              FROM tmp_asset_id_mapping
+              WHERE old_asset_id = asset_proofs.asset_id)
+  AND NOT EXISTS (SELECT 1
+                  FROM asset_proofs ap
+                  WHERE ap.asset_id = (SELECT new_asset_id
+                                       FROM tmp_asset_id_mapping
+                                       WHERE old_asset_id = asset_proofs.asset_id));
+
+-- Step 6: Remove duplicates from the asset_witnesses table.
+DELETE
+FROM asset_witnesses
+WHERE witness_id NOT IN (SELECT min(witness_id)
+                         FROM asset_witnesses
+                         GROUP BY asset_id, witness_index);
+
+-- Step 7: Re-enable the unique index on the asset_witnesses table.
+CREATE UNIQUE INDEX asset_witnesses_asset_id_witness_index_unique
+    ON asset_witnesses (
+                        asset_id, witness_index
+        );
+
+-- Step 8: Delete any duplicate proofs.
+DELETE
+FROM asset_proofs
+WHERE asset_id NOT IN (SELECT min_asset_id FROM tmp_min_assets);
+
+-- Step 9: Delete the duplicates from the assets table. This will then also
+-- delete dangling asset_witnesses, since we have a ON DELETE CASCADE clause on
```

Comment:
No, since we update all witnesses to point to the asset that we're going to keep, removing the duplicate assets won't remove the now duplicate witnesses. I guess I don't even have to mention DELETE CASCADE here...

Reply:
Would be good to remove that mention.